crates/starfish/core/src/transactions_synchronizer.rs

```diff
@@ -865,61 +894,59 @@ impl<C: NetworkClient, V: BlockVerifier, D: CoreThreadDispatcher>
                 Box::new(vec.into_iter())
             };
 
+        let mut assigned_block_refs_with_multiplicity: BTreeMap<BlockRef, usize> = BTreeMap::new();
+
         for (authority, authority_block_refs) in iter_authorities {
             let peer_hostname = &context.committee.authority(authority).hostname;
-            // Remove block_refs already assigned to another peer
-            let mut available_refs: BTreeSet<_> = authority_block_refs.clone();
-            // .difference(&assigned_block_refs)
-            // .cloned()
-            // .collect();
-            while !available_refs.is_empty() {
-                // TODO: remove this and use inflight_transactions_map to choose transactions
-                //  randomly
-                let selected_block_refs = available_refs
-                    .iter()
-                    .choose_multiple(&mut rng, context.parameters.max_transactions_per_fetch)
-                    .into_iter()
-                    .cloned()
-                    .collect::<BTreeSet<_>>();
 
-                if selected_block_refs.is_empty() {
-                    break;
-                }
+            // Filter: keep only blocks that have not reached the per-block limit
+            let selected_block_refs: BTreeSet<_> = authority_block_refs
+                .iter()
+                .filter(|b| {
+                    assigned_block_refs_with_multiplicity
+                        .get(*b)
+                        .copied()
+                        .unwrap_or(0)
+                        < MAX_ASSIGNMENTS_PER_BLOCK_REF_PER_RUN
+                })
+                .take(context.parameters.max_transactions_per_fetch)
```

Comment:
Here we will always take the first BlockRefs. Shouldn't we randomize it to avoid situation where the first elements are fetched from maximum number of peers while the last ones might not be fetched in extreme scenario?